openedx/core/djangoapps/programs/tasks.py

```diff
@@ -499,71 +508,93 @@ def _retry_with_custom_exception(username, course_run_key, reason, countdown):
     LOGGER.info(f"Running task award_course_certificate for user {user}")
     try:
         course_key = CourseKey.from_string(course_run_key)
-        # Get the cert for the course key and username if it's both passing and available in professional/verified
-        try:
-            certificate = GeneratedCertificate.eligible_certificates.get(
-                user=user.id,
-                course_id=course_key,
-            )
-        except GeneratedCertificate.DoesNotExist:
-            LOGGER.warning(
-                f"Task award_course_certificate was called for user {user} in course run {course_key} but this learner "
-                "has not earned a course certificate in this course run"
-            )
-            return
+    except InvalidKeyError as exc:
+        error_msg = "Failed to determine course key"
+        raise MaxRetriesExceededError(
+            f"Failed to award course certificate for user {username} for course {course_run_key}. Reason: {error_msg}"
+        ) from exc
+
+    # Get the cert for the course key and username if it's both passing and available in professional/verified
+    try:
+        certificate = GeneratedCertificate.eligible_certificates.get(
+            user=user.id,
+            course_id=course_key,
+        )
+    except GeneratedCertificate.DoesNotExist:
+        LOGGER.warning(
+            f"Task award_course_certificate was called for user {user} in course run {course_key} but this learner "
+            "has not earned a course certificate in this course run"
+        )
+        return
 
+    try:
         if (
-            certificate.mode in CourseMode.CERTIFICATE_RELEVANT_MODES
-            or CourseMode.is_eligible_for_certificate(certificate.mode)
+            certificate.mode not in CourseMode.CERTIFICATE_RELEVANT_MODES
+            and not CourseMode.is_eligible_for_certificate(certificate.mode)
         ):
-            course_overview = get_course_overview_or_none(course_key)
-            if not course_overview:
-                LOGGER.warning(
-                    f"Task award_course_certificate was called for user {user} in course {course_key} but no course "
-                    "overview could be retrieved for the course run"
-                )
-                return
+            return
```

Comment:
Should we also log a message here for debugging? Something along the lines of "Task {blah} was called for {user} in {course key} but the course run's mode is not eligible"?

Some wordsmithing may be required to try and make sense of it. Basically the situation here is... we (edx.org) used to award certificate to learners in audit and honor courses. We have stopped that practice long ago, but there still remains some (legitimate) records that we don't want to pass on to Credentials. 

Certificates in the honor/audit mode are "not interesting" to Credentials.

Reply:
Maybe not as big of a priority since we are specifically using a non-default ORM manager to only retrieve "eligible certificates".

Reply:
EligibleCertificateManager: https://github.com/openedx/edx-platform/blob/c5967361176eb2cefb75bc2316e0ecb7a64a7cc5/lms/djangoapps/certificates/models.py#L119-L138